packages/headless/src/controllers/pager/headless-pager.ts

```diff
@@ -141,6 +142,34 @@ export function buildPager(
   engine: SearchEngine,
   props: PagerProps = {}
 ): Pager {
+  const {dispatch} = engine;
+  const pager = buildCorePager(engine, props);
+
+  return {
+    ...pager,
+
+    get state() {
+      return pager.state;
+    },
+
+    selectPage(page: number) {
+      pager.selectPage(page);
+      dispatch(executeSearch(logPageNumber()));
+    },
+
+    nextPage() {
+      pager.nextPage();
+      dispatch(executeSearch(logPageNext()));
+    },
+
+    previousPage() {
+      pager.previousPage();
+      dispatch(executeSearch(logPagePrevious()));
+    },
+  };
+}
```

Comment:
@francoislg, @dlafreniere in the case of the listing api, with this approach, you would need to duplicate lines 141-170, adjust the engine type on line 142, and dispatch a dedicated action to get the listing page on lines 157, 162, 167.

In the pagination reducer, you would need to add a [new case to handle](https://github.com/coveo/ui-kit/blob/master/packages/headless/src/features/pagination/pagination-slice.ts#L89) the listing page action. The only piece of info needed from the response is `totalCountFiltered` (aka total number of results) to determine the max page.

Reply:
Looks good! Seems to make it quite easy to extend :P

Reply:
I want to challenge a bit on what we are actually gaining here.

We will have some code repetition removed (although it does not seems quite a lot), and the code repetition we remove seems pretty "trivial code" at the end of the day ?

So by doing DRY, it seems we get a bit more complex code (inheritance-ish like) for not much net gain. Because the code is not complex, the chance of having bugs that we need to solve in both pager function is relatively small.

If the thing we want to solve is that we would like both pager (search and listing API) to return the same object shape, could we instead extract an interface ?

And then we live with the fact that there's going to be some repetition for both pager, but I feel the code might be more simple in the end.

What's your opinion on this ?

Reply:
> If the thing we want to solve is ...

The problem we are trying to solve is that certain controllers such as the `Pager` are coupled to the search api, making them unusable by other apis. The coupling occurs in two ways:

1. Whether a controller is dispatching api-specific async thunk actions.
2. Whether a controller is reading from the `search` state property (not the case with the `Pager`)

**How to overcome this?**

One approach can be to duplicate the controller and change a few lines. The duplication is ~80 lines of source code, and ~180 lines of unit test code, so ~ 260 lines. Given `n` apis, this scales at `260n`. If there is a bug or change in the core logic, it would need to be fixed (and ideally unit-tested) in `n` places.

I think minimizing repetition will make it easier to maintain the code-base long-term, and ensure the core logic stays consistent across use-cases.

Reply:
Out of curiosity could we have done it such that the logic of the pager be common/fixed and that the SearchEngine and (eventually) ListingPageEngine share a common interface "executeSearch"? in the SearchEngine it would call the search API + log analytics, and in the ListingPageEngine it would call the listingPage API?

Reply:
This could have been another approach, but I found it to be not as clean because it requires passing props to the common function according to the common interface. Since the analytics action differs with each `executeSearch`, three different cases need to be handled. This is not the case with the listing pager though. e.g.
```
function buildPager(engine, props) {
    const executeSearchOnNextPage = executeSearch(logNextPage());
    const executeSearchOnPreviousPage = executeSearch(logPreviousPage());
    ...

    return buildCorePager(engine, props, {executeSearchOnNextPage, executeSearchOnPreviousPage, ...})
}

function buildListPager() {
    const executeSearchOnNextPage = fetchListingPage()
    const executeSearchOnPreviousPage = fetchListingPage();
    ...

    return buildCorePager(engine, props, {executeSearchOnNextPage, executeSearchOnPreviousPage, ...})
}


```